### PR TITLE
v4.0.0

### PR DESCRIPTION
**Pull request by Pro7ech**
Updates README.md and CHANGELOG.md for the v4.0.0 release.

Also fixe various bugs discovered in the mean time.

### CODE REVIEW

**Review by jrtroncoso — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -65,29 +65,47 @@ on Lattigo:
...
+
+External pull requests should only be used to propose new functionalities that are substantial and would require a fair amount of work if done on our side. If you plan to open such a pull request, please contact us before doing so to make sure that the proposed changes are
+aligned with our development roadmap.
+
+External pull requests only proposing small or trivial changes will be converted to an issue and closed.
+
+## Support and Issues
+
+The GitHub issues should be only used for bug reports and questions directly related to the use or the implementation of the library.
+Any other issue will be closed, and for this we recommend the use of GitHub discussions or other topic-specific forums instead.
```
Comment (romainbou) [inline README.md:88]:
```suggestion
Any other issue will be closed, and for this we recommend the use of [GitHub discussions](https://github.com/tuneinsight/lattigo/discussions) or other topic-specific forums instead.
```

File: README.md [hunk 2]
```diff
@@ -65,29 +65,47 @@ on Lattigo:
...
+Any new issue regarding an unexpected behavior of the library or one of its packages must be accompanied 
+by a self-contained `main.go` reproducing the unwanted behavior.
+
 ## License
 
 Lattigo is licensed under the Apache 2.0 License. See LICENSE.
```
Comment (romainbou) [inline README.md:94]:
```suggestion
Lattigo is licensed under the Apache 2.0 License. See [LICENSE](https://github.com/tuneinsight/lattigo/blob/master/LICENSE).
```

File: README.md [hunk 3]
```diff
@@ -65,29 +65,47 @@ on Lattigo:
...
- 
+fixes and new features. Hence, we encourage all Lattigo users to update to the latest Lattigo version.
+
 
 See CHANGELOG.md for the current and past versions.
 
+## Stability
+
+To keep a comprehensive history, we prioritize rebases over merges for branches other than `main`.
+Branches with the prefix `dev_` are branches in active development and will be frequently rebased.
```
Comment (romainbou) [inline README.md:76]:
```suggestion
Branches with the prefix `dev_` are branches in active development and will be frequently rebased. Hence, we don't recommend depending on them.
```

**Review by jrtroncoso — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -65,29 +65,47 @@ on Lattigo:
...
+
 ## License
 
-Lattigo is licensed under the Apache 2.0 License. See LICENSE.
+Lattigo is licensed under the Apache 2.0 License. See [LICENSE](https://github.com/tuneinsight/lattigo/blob/master/LICENSE).
 
 ## Contact
 
-If you want to contribute to Lattigo or you have any suggestion, do not hesitate to contact us at
+If you want to contribute to Lattigo or you have any questions or suggestions, do not hesitate to contact us at
```
Comment (ChristianMct) [inline README.md:98]: Shouldn't questions and suggestions go on the public discussion by default ? Should we say "non-public inquires" or something like that ?

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,7 +2,7 @@
 # Changelog
 All notable changes to this library are documented in this file. 
 
-# [3.1.0] - UNREALEASED
+# [4.0.0]
```
Comment (ChristianMct) [inline CHANGELOG.md:5]: Bumping the major version number requires package name change to `github.com/tuneinsight/lattigo/v4` (along with associated changes).

**Review by ChristianMct — CHANGES_REQUESTED** [review]

**Comment by jrtroncoso** [comment]
Should (all or some of) the functions in bfv/utils and bgv/utils be moved to rlwe/utils and made generic? There's no specific BFV/BGV function there besides bgv.Norm and bfv.Noise, which are used for the same purpose while indeed assuming a non-generic ciphertext structure (this could be generalized, e.g., by specifying the noise scale and modulus as input) despite getting rlwe generic types as inputs. Also, errorStats is replicated in bgv/decryptor and in bfv/utils, but only called in tests from the Norm and Noise functions respectively. As this is where BGV is merged, this could be considered already here to homogenize the noise evaluation of integer-based encryptions. Alternatively, it could be added as a separate minor PR later on.

**Comment by Pro7ech** [comment]
I'm actively working on moving a lot of generic function from all schemes to the rlwe package. If you are ok with it, we can delay it to one of the PR that will tackle theses changes.

**Review by jrtroncoso — CHANGES_REQUESTED** [review]
Comments:
- bgv/evaluator: Automorphism and AutomorphismHoisted lack documentation
- ckks/bootstrapping/bootstrap_bench_test: for consistency, rename "*Bootstrapp" to "*Bootstrap"

**Review by jrtroncoso — APPROVED** [review]

**Review by ChristianMct — APPROVED** [review]

**Review by romainbou — APPROVED** [review]

**Pull request merged**